### PR TITLE
C1 khloud c3 s3 register as developer

### PR DESCRIPTION
**Pull request by Khloud-Khalid**
issue #85 

UML reviewer @Mirna-Yacout 
Implementation reviewers @KareemAliRahman  , @adamggg 
Documentation reviewer @zoheiry

### CODE REVIEW

**Comment by Mirna-Yacout** [comment]
@Khloud-Khalid UML on the computer would be clearer

File: arability/app/controllers/developer_controller.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# description:
+#  	generates view with form for registration as developer
+# params:
+#  	none
+# success:
+#  	view generated successfully
+# failure:
+# 	gamer not signed in  	
+	def new
+		if(gamer_signed_in?)
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: You shouldn't use the gamer_signed_in method inside a controller action but instead you would need to use a before_filter :authenticate_gamer for the whole controller

File: arability/app/controllers/developer_controller.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#  	first_name, last_name
+# success:
+#  	developer created successfully
+# failure:
+# 	invalid information, user already registered as developer
+	def create
+		@developer = Developer.new(params[:developer])
+		@developer.gamer_id = current_gamer.id
+		if((Developer.find_by_gamer_id(current_gamer.id)) == nil)
+			if @developer.save
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: You're not sending any notice on success? You should fix this.

File: arability/app/controllers/developer_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+		@developer.gamer_id = current_gamer.id
+		if((Developer.find_by_gamer_id(current_gamer.id)) == nil)
+			if @developer.save 
+				render '/my_subscriptions/new'
+			else
+				flash[:notice] = "Failed to complete registration."
+			end
+		else
+			#redirect to homepage
+			flash[:notice] = "You are already registered as developer."
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: You're not redirecting after you set the notice, you need to render.

File: arability/app/controllers/my_subscription_controller.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+# description:
+#  	generates view with form for choosing subscription model
+# params:
+#  	none
+# success:
+#  	view generated successfully
+# failure:
+# 	gamer not signed in  
+	def new
+		if(gamer_signed_in?)
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: You need to use the before_filter instead of authenticating per action.

File: arability/app/controllers/my_subscription_controller.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+			@my_subscription = MySubscription.new
+		else
+			flash[:notice] = "Please login to proceed with registration."
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: Again you're just setting the notice and not rendering anything.

File: arability/app/controllers/my_subscription_controller.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+# failure:
+# 	invalid information
+	def create
+		@subscription_model = params[:subscription_model]
+		@my_subscription = MySubscription.new()
+		@my_subscription.developer_id = current_developer.id
+		@my_subscription.subscription_models_id = @subscription_model
+		if @my_subscription.save 
+			#redirect to home page
+			flash[:notice] = "You have successfully registered as a developer."
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: Where is the render?

File: arability/app/controllers/my_subscription_controller.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+		else
+			flash[:notice] = "Failed to complete registration."
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: Where is the render?

**Comment by adamggg** [comment]
@Khloud-Khalid, Code reopened, and check the comments I have added on your files that are changed in the files changed tab in the pull request.

File: arability/app/assets/javascripts/application.js [hunk 8]
```diff
@@ -13,4 +13,3 @@
 //= require jquery
 //= require jquery_ujs
 //= require bootstrap
-//= require_tree .
```
Comment (adamggg) [inline arability/app/assets/javascripts/application.js:16]: Why did you remove this?

File: arability/app/controllers/my_subscription_controller.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+		else
+			flash[:notice] = "Please login to proceed with registration."
+		end	
+	end
+# author:
+# 	Khloud Khalid
+# description:
+#  	creates new developer using parameters from registration form
+# params:
+#  	first_name, last_name
```
Comment (zoheiry) [inline arability/app/controllers/my_subscription_controller.rb]: @Khloud-Khalid  This is not the correct documentation for this fucntion, i cant see any params called first_name or last_name and this method creates a subscription model not a new developer

File: arability/app/models/developer.rb [hunk 10]
```diff
@@ -1,4 +1,7 @@
 class Developer < ActiveRecord::Base
   belongs_to :gamer
-  # attr_accessible :title, :body
+   attr_accessible :first_name, :last_name, :verified, :gamer_id
+   validates :first_name,  :presence => true
+   validates :last_name,  :presence => true
+   validates :gamer_id, :presence => true
 end
```
Comment (krmali) [inline arability/app/models/developer.rb:13]: You can add more reliable validations here , as validating the uniqueness of the gamer_id , so no one gamer has more than one developer accounts.

**Comment by Khloud-Khalid** [comment]
@Mirna-Yacout , @adamggg , @KareemAliRahman , @zoheiry  please review now

**Comment by Mirna-Yacout** [comment]
@Khloud-Khalid please check ur if condition, an if condition should be put in a box separating the if and else parts with a dotted line (refer to the lecture for more details)

File: arability/app/models/developer.rb [hunk 11]
```diff
@@ -1,4 +1,7 @@
 class Developer < ActiveRecord::Base
   belongs_to :gamer
-  # attr_accessible :title, :body
+   attr_accessible :first_name, :last_name, :verified, :gamer_id
+   validates :first_name,  :presence => true
```
Comment (adamggg) [inline arability/app/models/developer.rb]: You need to add regex validation to make sure that the first and last name has no spaces.

File: arability/app/controllers/developer_controller.rb [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+# params:
+#  	first_name, last_name
+# success:
+#  	developer created successfully
+# failure:
+# 	invalid information, user already registered as developer
+	def create
+		@developer = Developer.new(params[:developer])
+		@developer.gamer_id = current_gamer.id
+		if @developer.valid?
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: By adding the validations to the actual model you wouldn't need to do this here since @developer.save will check the validations before saving and will return true if it was successful

File: arability/app/views/my_subscription/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,9 @@
+<h2>Register as developer Step2: choose your subscription model</h2>
+
+<%= form_for(:my_subscription, :url => { :controller => "my_subscription",:action => "create" }) do |f| %>
+<% SubscriptionModel.find(:all).each do |item| %>
+  <p><%item.name%></p><%= radio_button_tag(:my_subscription, item.id) %>
```
Comment (adamggg) [inline arability/app/views/my_subscription/new.html.erb]: The item.name isn't printing in the loop because you didn't put an = after <% and please change the item name to a more descriptive variable name and no need for SubscriptionModel.find(:all) you can just do SubscriptionModel.all

File: arability/app/views/my_subscription/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+<% end %>
+  <div><%= f.submit "Register" %></div>
```
Comment (adamggg) [inline arability/app/views/my_subscription/new.html.erb]: This shouldn't be called register but instead choose plan or something more relevant.

File: arability/app/controllers/my_subscription_controller.rb [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+# description:
+#  	creates new my_subscription using parameters from registration form and links it to the developer
+# params:
+#  	subscription_model_id
+# success:
+#  	my_subscription created successfully and linked to developer
+# failure:
+# 	invalid information
+	def create
+		@my_subscription = MySubscription.new()
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: No need for the () at the end

**Comment by Khloud-Khalid** [comment]
@Mirna-Yacout , @adamggg  please review

File: arability/app/views/my_subscription/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+<h2>Register as developer Step2: choose your subscription model</h2>
+
+<%= form_for(:my_subscription, :url => { :controller => "my_subscription",:action => "create" }) do |f| %>
+<% SubscriptionModel.all.each do |subscription| %>
+  <p><%=subscription.name%></p><%= radio_button_tag(:my_subscription, subscription.id) %>
```
Comment (adamggg) [inline arability/app/views/my_subscription/new.html.erb]: <%= code %> , you have to add a space before the subscription.name nor after it.

File: arability/app/models/developer.rb [hunk 17]
```diff
@@ -1,4 +1,9 @@
...
+   validates :last_name,  :presence => true
+   validates_format_of :first_name, :with => /^[a-zA-Z0-9][a-zA-Z0-9_]*$/
```
Comment (adamggg) [inline arability/app/models/developer.rb]: This regular expression allows for numerals as part of the name and add a max length and min length validation as well on both first and last name.

File: arability/app/controllers/my_subscription_controller.rb [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+# success:
+#  	my_subscription created successfully and linked to developer
+# failure:
+# 	invalid information
+	def create
+		@my_subscription = MySubscription.new
+		@my_subscription.developer_id = Developer.find_by_gamer_id(3).id
+		@my_subscription.subscription_models_id = params[:my_subscription]
+		if @my_subscription.save 
+			flash[:notice] = "You have successfully registered as a developer."
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: You should render the developer back to home page but this will be changed later where will most probably redirect him to the developer page so for now just have it render the home page.

**Comment by adamggg** [comment]
@Khloud-Khalid, code reopened, check the comments on the files.

**Comment by Mirna-Yacout** [comment]
@Khloud-Khalid dotted line not dashes in the if condition

File: arability/app/controllers/developer_controller.rb (hunk 12 above)
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: Why are you still using this? The developer is checked to be valid upon .save which runs the model validations.

**Comment by Mirna-Yacout** [comment]
reviewed

File: arability/app/models/developer.rb [hunk 19]
```diff
@@ -1,4 +1,13 @@
 class Developer < ActiveRecord::Base
   belongs_to :gamer
-  # attr_accessible :title, :body
+   attr_accessible :first_name, :last_name, :verified, :gamer_id
+   validates :first_name,  :presence => true
+   validates :last_name,  :presence => true
+   validates_format_of :first_name, :with => /^[a-zA][a-zA_]*$/
```
Comment (adamggg) [inline arability/app/models/developer.rb]: These regular expression is incorrect, test them out, they don't even accept for first name: Adam and last name Ghanem it gives the image I have attached

Screen shot 2013-03-29 at 6 02 11 PM

File: arability/app/models/developer.rb [hunk 20]
```diff
@@ -1,4 +1,13 @@
...
+   validates_format_of :last_name, :with => /^[a-zA][a-zA-_]*$/
+   validates_length_of :first_name, :maximum => 12
```
Comment (adamggg) [inline arability/app/models/developer.rb]: Let the max length be 18 instead of 12 since some Arabic names are pretty long.

**Comment by adamggg** [comment]
Check the comments I have added inline to the code and also the image.

File: arability/app/controllers/my_subscription_controller.rb [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+# params:
+#  	subscription_model_id
+# success:
+#  	my_subscription created successfully and linked to developer
+# failure:
+# 	invalid information
+	def create
+		@my_subscription = MySubscription.new
+		@my_subscription.developer_id = Developer.find_by_gamer_id(current_gamer.id).id
+		@my_subscription.subscription_models_id = params[:my_subscription]
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: Check if there exists a subscription model by the id passed in by the request to make sure that the user doesn't mess around with changing the header variables and entering not dangerous but unwanted nor true data and also will cause issues later when we query the model related to the developer.

File: arability/app/views/my_subscription/new.html.erb [hunk 22]
```diff
@@ -0,0 +1,9 @@
+<h2>Register as developer Step2: choose your subscription model</h2>
+
+<%= form_for(:my_subscription, :url => { :controller => "my_subscription",:action => "create" }) do |f| %>
+<% SubscriptionModel.all.each do |subscription| %>
```
Comment (adamggg) [inline arability/app/views/my_subscription/new.html.erb:3]: I added a comment in the create action in the subscription controller to make sure that user doesn't mess around.

**Comment by adamggg** [comment]
Check the comments I have added and we're almost done I think =)

File: arability/app/controllers/developer_controller.rb (hunk 12 above)
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: @Khloud-Khalid, this is the third time I comment on this line. I think you get the picture. 
Screen shot 2013-03-29 at 6 32 37 PM

**Comment by adamggg** [comment]
@Khloud-Khalid, everything seems to be working just fine but when I don't choose any subscription model I get the same error message. Change it to something more descriptive like "Please choose a subscription model to continue"

Screen shot 2013-03-29 at 6 37 59 PM

And also please fix the @developer.valid? thingy cause it seems very redundant since we already use @developer.save which runs all the validations.

File: arability/app/controllers/developer_controller.rb [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+# params:
+#   first_name, last_name
+# success:
+#   developer created successfully
+# failure:
+#   invalid information, user already registered as developer
+def create
+  @developer = Developer.new(params[:developer])
+  @developer.gamer_id = current_gamer.id
+  if @developer.valid?
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: How many times do I have to say this really? Why are you using .valid? .save checks for validations and you are already using .save

File: arability/app/controllers/my_subscription_controller.rb [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+#   invalid information
+def create
+  if(SubscriptionModel.find_by_id(params[:my_subscription]) != nil)
+    @my_subscription = MySubscription.new
+    @my_subscription.developer_id = Developer.find_by_gamer_id(current_gamer.id).id
+    @my_subscription.subscription_models_id = params[:my_subscription]
+    if @my_subscription.save
+      flash[:notice] = "You have successfully registered as a developer."
+      render 'pages/home'
+      else
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: Incorrect indentation here.

File: arability/app/controllers/developer_controller.rb [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+def create
+  @developer = Developer.new(params[:developer])
+  @developer.gamer_id = current_gamer.id
+  if @developer.save
+    render 'my_subscription/new'
+    else
+      if((Developer.find_by_gamer_id(current_gamer.id)) != nil)
+        flash[:notice] = "You are already registered as a developer."
+        render :action => 'new'
+        else
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: Incorrect indentation here too.

**Comment by adamggg** [comment]
@Khloud-Khalid, still some indentation convention problems.

File: arability/app/controllers/developer_controller.rb [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+#   Khloud Khalid
+# description:
+#   creates new developer using parameters from registration form and renders my_subscription#new
+# params:
+#   first_name, last_name
+# success:
+#   developer created successfully
+# failure:
+#   invalid information, user already registered as developer
+def create
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: Why are the methods on the same indentation level as the class? you need to shift every single line >>

File: arability/app/controllers/my_subscription_controller.rb [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+    @my_subscription.developer_id = Developer.find_by_gamer_id(current_gamer.id).id
+    @my_subscription.subscription_models_id = params[:my_subscription]
+    if @my_subscription.save
+      flash[:notice] = "You have successfully registered as a developer."
+      render 'pages/home'
+    else
+      flash[:notice] = "Failed to complete registration."
+      render 'my_subscription/new'
+    end
+  else
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: To what does this else belong to? Fix the indentation.

File: arability/app/models/developer.rb [hunk 28]
```diff
@@ -1,4 +1,13 @@
 class Developer < ActiveRecord::Base
-  belongs_to :gamer
-  # attr_accessible :title, :body
+ belongs_to :gamer
+ attr_accessible :first_name, :last_name, :verified, :gamer_id
+ validates :first_name,  :presence => true
```
Comment (adamggg) [inline arability/app/models/developer.rb]: Why is there 2 white spaces between , and :presence?

File: arability/app/models/my_subscription.rb [hunk 29]
```diff
@@ -1,4 +1,5 @@
 class MySubscription < ActiveRecord::Base
-  attr_accessible :developer, :word_add, :word_follow, :word_search
+  attr_accessible :developer, :word_add, :word_follow, :word_search, :subscription_models_id
   belongs_to :developer
+  validates :subscription_models_id,  :presence => true
```
Comment (adamggg) [inline arability/app/models/my_subscription.rb]: extra whitespace.

File: arability/app/controllers/my_subscription_controller.rb [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+# description:
+#   creates new my_subscription using parameters from registration form and links it to the developer
+# params:
+#   subscription_model_id
+# success:
+#   my_subscription created successfully and linked to developer
+# failure:
+#   invalid information
+  def create
+    if(SubscriptionModel.find_by_id(params[:my_subscription]) != nil)
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb]: No need for the () in the if condition, do this in all if conditions also.

File: arability/app/controllers/my_subscription_controller.rb [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+#   Khloud Khalid
+# description:
+#   generates view with form for choosing subscription model
+# params:
+#   none
+# success:
+#   view generated successfully
+# failure:
+#   gamer not signed in  
+  def new
```
Comment (adamggg) [inline arability/app/controllers/my_subscription_controller.rb:13]: I revisited /my_subcriptions/new and I was able to choose my subscription model again. This shouldn't be allowed. Add a check to the create and new action to make sure that he can't choose his subscription model again. Add this check to the new and create action as well. Please.

File: arability/app/controllers/developer_controller.rb [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+#   Khloud Khalid
+# description:
+#   generates view with form for registration as developer
+# params:
+#   none
+# success:
+#   view generated successfully
+# failure:
+#   gamer not signed in
+  def new
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb:13]: Why can I still visit the page where I can register as a developer if I am already a developer? Add the check to both the create and new. Please. :frowning:

File: arability/app/controllers/developer_controller.rb [hunk 33]
```diff
@@ -0,0 +1,48 @@
...
+     if Developer.find_by_gamer_id(current_gamer.id) != nil
+        flash[:notice] = "You are already registered as a developer. Don't you remember?"
+        render :action => 'new'
```
Comment (adamggg) [inline arability/app/controllers/developer_controller.rb]: You're still rendering the signup as a developer page. Render it to the home page for now :smile:

**Comment by adamggg** [comment]
I can still visit the page where I can choose my subscription model even after choosing it.

**Comment by adamggg** [comment]
@Khloud-Khalid, by visiting http://0.0.0.0:3000/developers/create and http://0.0.0.0:3000/my_subscriptions/create after you have subscribed as a developer still is visitable you should check for that and also check that I can't visit the page where I can pick my subscription before I register as a developer as well =)

Hanit!!

**Comment by adamggg** [comment]
You fixed it for not being able to register as a developer again but you still haven't made for the path /my_subscriptions/create it works for the /new though. Almost done!! So close!

**Comment by adamggg** [comment]
Fix conflicts and tell me. Everything looks el7amdulilah =)

Rabena yestor!
